src/repo/kv.rs

```diff
@@ -101,234 +95,301 @@ impl DataStore for KvDataStore {
-
-        let mut batch = sled::Batch::default();
+        use ConflictableTransactionError::Abort;
+        let res = self.get_db().transaction(|tx_tree| {
+            let already_pinned = get_pinned_mode(&tx_tree, target)?;
 
-        match already_pinned {
-            Some(PinMode::Direct) => return Ok(()),
-            Some(PinMode::Recursive) => return Err(anyhow::anyhow!("already pinned recursively")),
-            Some(PinMode::Indirect) => {
-                let pin_key = get_pin_key(target, &PinMode::Indirect);
-                batch.remove(pin_key.as_str());
+            match already_pinned {
+                Some(PinMode::Direct) => return Ok(false),
+                Some(PinMode::Recursive) => {
+                    return Err(Abort(anyhow::anyhow!("already pinned recursively")))
+                }
+                Some(PinMode::Indirect) => {
+                    // TODO: I think the direct should live alongside the indirect?
+                    let pin_key = get_pin_key(target, &PinMode::Indirect);
+                    tx_tree.remove(pin_key.as_str())?;
+                }
+                None => {}
             }
-            None => {}
-        }
 
-        let direct_key = get_pin_key(target, &PinMode::Direct);
+            let direct_key = get_pin_key(target, &PinMode::Direct);
+            tx_tree.insert(direct_key.as_str(), "")?;
 
-        batch.insert(direct_key.as_str(), "");
+            Ok(true)
+        });
 
-        Ok(self._apply_batch(batch)?)
+        if launder(res)? {
+            self.flush_async().await
+        } else {
+            Ok(())
+        }
     }
 
     async fn insert_recursive_pin(
         &self,
         target: &Cid,
         referenced: References<'_>,
     ) -> Result<(), Error> {
+        // since the transaction can be retried multiple times, we need to collect these and keep
+        // iterating it until there is no conflict.
         let set = referenced.try_collect::<BTreeSet<_>>().await?;
 
-        let mut batch = sled::Batch::default();
-        let already_pinned = get_pinned_mode(self, target)?;
-
-        match already_pinned {
-            Some(PinMode::Recursive) => return Ok(()),
-            Some(mode @ PinMode::Direct) | Some(mode @ PinMode::Indirect) => {
-                let key = get_pin_key(target, &mode);
-                batch.remove(key.as_str());
-            }
-            None => {}
-        }
+        // the transaction is not infallible but there is no additional error we return
+        let modified = self
+            .get_db()
+            .transaction::<_, _, Infallible>(move |tx_tree| {
+                let already_pinned = get_pinned_mode(tx_tree, target)?;
+
+                match already_pinned {
+                    Some(PinMode::Recursive) => return Ok(false),
+                    Some(mode @ PinMode::Direct) | Some(mode @ PinMode::Indirect) => {
+                        // FIXME: this is probably another lapse in tests that both direct and
```

Comment:
Whoops, forgot to reply to this one. Yeah, this is a good point.

I've implemented (or at least tried to) the mem and fs so that the pins would not replace other pins, and I am surprised you found this in go-ipfs. I might also misremember, I did go back and forth here a lot. But I guess I'll need to confirm my thinking in upcoming PRs with more test cases, following the addition of `spawn_blocking`.